src/libraries/System.Text.Json/gen/JsonSourceGenerator.Parser.cs

```diff
@@ -81,109 +84,286 @@ public Parser(Compilation compilation)
+                    contextGenSpecList.Add(new ContextGenerationSpec
+                    {
+                        SerializerOptions = options ?? new JsonSerializerOptionsAttribute(),
+                        ContextType = contextTypeSymbol.AsType(_metadataLoadContext),
+                        RootSerializableTypes = rootTypes,
+                    });
+
+                    // Clear the cache of generated metadata between the processing of context classes.
+                    _typeGenerationSpecCache.Clear();
+                }
+
+                if (contextGenSpecList == null)
+                {
+                    return null;
+                }
+
+                return new SourceGenerationSpec
+                {
+                    ContextGenerationSpecList = contextGenSpecList,
+                    BooleanType = _booleanType,
+                    ByteArrayType = _byteArrayType,
+                    CharType = _charType,
+                    DateTimeType = _dateTimeType,
+                    DateTimeOffsetType = _dateTimeOffsetType,
+                    GuidType = _guidType,
+                    StringType = _stringType,
+                    NumberTypes = _numberTypes,
+                };
+            }
+
+            // Returns true if a given type
+            // - derives from JsonSerializerContext
+            // - is partial
+            // - has no members
+            private bool IsConfigurableJsonSerializerContext(
+                ClassDeclarationSyntax classDeclarationSyntax,
+                INamedTypeSymbol jsonSerializerContextSymbol,
+                SemanticModel compilationSemanticModel)
+            {
+                SeparatedSyntaxList<BaseTypeSyntax>? baseTypeSyntaxList = classDeclarationSyntax.BaseList?.Types;
+                if (baseTypeSyntaxList == null)
+                {
+                    return false;
+                }
+
+                INamedTypeSymbol? match = null;
+
+                foreach (BaseTypeSyntax baseTypeSyntax in baseTypeSyntaxList)
+                {
+                    INamedTypeSymbol? candidate = compilationSemanticModel.GetSymbolInfo(baseTypeSyntax.Type).Symbol as INamedTypeSymbol;
+                    if (candidate != null && jsonSerializerContextSymbol.Equals(candidate, SymbolEqualityComparer.Default))
+                    {
+                        match = candidate;
+                        break;
+                    }
+                }
+
+                if (match == null)
+                {
+                    return false;
+                }
+
+                bool isPartial = classDeclarationSyntax.Modifiers.Where(token => token.IsKind(Microsoft.CodeAnalysis.CSharp.SyntaxKind.PartialKeyword)).FirstOrDefault() != default;
+                if (!isPartial)
+                {
+                    return false;
+                }
 
-                        if (typeSymbol == null)
+                return classDeclarationSyntax.Members.Count() == 0;
```

Comment:
I kind of think this rule is a bit prohibitive. Why wouldn't developers want to be able to define their own "helper" methods on their JsonContext class?

I feel like the rule should be:
            // - derives from JsonSerializerContext (directly or indirectly)
            // - has at least one `JsonSerializable` attribute
            // - is partial

Reply:
I've removed the requirement that it has no members. I assume the expectation is that if there are clashes between members (user authored vs generated), the onus would be on the user to resolve them, e.g. renaming a member name.

Currently we support the context class deriving directly from `JsonSerializerContext`. I want to defer support for deriving indirectly to a follow up focused on scenarios such as:

```cs
[JsonSerializable(typeof(JsonMessage))]
internal partial class JsonContext : JsonSerializerContext { }

[JsonSerializable(typeof(Point))]
internal partial class AnotherContext : JsonContext { }
```

- Should the[`GetTypeInfo`](https://github.com/dotnet/runtime/blob/07d2439dd251006d2a3acb0e940b46076ba895cc/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerContext.cs#L107) override for `AnotherContext` contain metadata for the `JsonMessage` type?
- Seems we'd have to keep track of what was generated for `JsonContext` so there are no conflicts in the generated source.

Reply:
Made some changes in https://github.com/dotnet/runtime/pull/53212/commits/2dbebc9df14dbc7c966be22f6a75db5384af8829 that align better with https://github.com/dotnet/runtime/pull/53212#discussion_r639231466.

Reply:
I agree that we can defer deriving indirectly from `JsonSerializerContext`.